File: python/futoshiki.py
VERBOSE_LIMIT = 10

def parse_input(lines):
    """Parse the 15-line input format into horizontal and vertical constraints."""
    horizontal = [[None] * 6 for _ in range(9)]  # 9 rows, 6 possible horizontal constraints each
    vertical = [[None] * 9 for _ in range(6)]    # 6 constraint rows, 9 columns each
    
    h_idx = 0  # Index for horizontal constraint rows
    v_idx = 0  # Index for vertical constraint rows
    
    for i, line in enumerate(lines):
        # Get all constraint symbols in this line
        symbols = [c for c in line if c in '<>^v']
        horizontal_symbols = [c for c in symbols if c in '<>']
        vertical_symbols = [c for c in symbols if c in '^v']
        
        if horizontal_symbols and not vertical_symbols:
            # This is a horizontal constraint line
            if h_idx < 9:
                col_idx = 0
                for c in line:
                    if c in '<>':
                        if col_idx < 6:
                            horizontal[h_idx][col_idx] = c
                            col_idx += 1
                h_idx += 1
        elif vertical_symbols and not horizontal_symbols:
            # This is a vertical constraint line
            if v_idx < 6:
                col_idx = 0
                for c in line:
                    if c in '^v':
                        if col_idx < 9:
                            vertical[v_idx][col_idx] = c
                            col_idx += 1
                v_idx += 1
    
    return horizontal, vertical


def is_valid(grid, row, col, val, horizontal, vertical):
    """Check if placing val at (row, col) is valid."""
    # Check row constraint
    for c in range(9):
        if grid[row][c] == val:
            return False
    
    # Check column constraint
    for r in range(9):
        if grid[r][col] == val:
            return False
    
    # Check 3x3 subgrid constraint
    start_row, start_col = (row // 3) * 3, (col // 3) * 3
    for r in range(start_row, start_row + 3):
        for c in range(start_col, start_col + 3):
            if grid[r][c] == val:
                return False
    
    # Check inequality constraints
    # Horizontal constraints (between columns)
    constraint_positions = [0, 1, 3, 4, 6, 7]  # Skip positions 2 and 5 (thick lines)
    
    for i, pos in enumerate(constraint_positions):
        if col == pos and col + 1 < 9:
            # Check constraint between col and col+1
            constraint = horizontal[row][i]
            if constraint and grid[row][col + 1] != 0:
                if constraint == '<' and val >= grid[row][col + 1]:
                    return False
                elif constraint == '>' and val <= grid[row][col + 1]:
                    return False
        elif col == pos + 1 and col - 1 >= 0:
            # Check constraint between col-1 and col
            constraint = horizontal[row][i]
            if constraint and grid[row][col - 1] != 0:
                if constraint == '<' and grid[row][col - 1] >= val:
                    return False
                elif constraint == '>' and grid[row][col - 1] <= val:
                    return False
    
    # Vertical constraints (between rows)
    constraint_rows = [0, 1, 3, 4, 6, 7]  # Skip rows 2 and 5 (thick lines)
    
    for i, pos in enumerate(constraint_rows):
        if row == pos and row + 1 < 9:
            # Check constraint between row and row+1
            constraint = vertical[i][col]
            if constraint and grid[row + 1][col] != 0:
                if constraint == '^' and val >= grid[row + 1][col]:
                    return False
                elif constraint == 'v' and val <= grid[row + 1][col]:
                    return False
        elif row == pos + 1 and row - 1 >= 0:
            # Check constraint between row-1 and row
            constraint = vertical[i][col]
            if constraint and grid[row - 1][col] != 0:
                if constraint == '^' and grid[row - 1][col] >= val:
                    return False
                elif constraint == 'v' and grid[row - 1][col] <= val:
                    return False
    
    return True


def solve(grid, horizontal, vertical, verbose=False):
    """Solve the enhanced Sudoku using backtracking."""
    step_count = 0
    
    def find_empty():
        for r in range(9):
            for c in range(9):
                if grid[r][c] == 0:
                    return r, c
        return None, None
    
    def solve_recursive():
        nonlocal step_count
        step_count += 1
        
        if verbose and step_count <= VERBOSE_LIMIT:
            print(f"Step {step_count}: Looking for empty position...")
        
        row, col = find_empty()
        if row is None:
            return True  # All positions filled
        
        for val in range(1, 10):
            if is_valid(grid, row, col, val, horizontal, vertical):
                grid[row][col] = val
                
                if verbose and step_count <= VERBOSE_LIMIT:
                    print(f"  Placed {val} at ({row}, {col})")
                
                if solve_recursive():
                    return True
                
                grid[row][col] = 0  # Backtrack
        
        return False
    
    return solve_recursive()


def print_solution(grid):
    """Print the solution in the required format."""
    for row in grid:
        print(' '.join(map(str, row)))


def solve_enhanced_sudoku(input_text, verbose=False):
    """Main function to solve enhanced Sudoku from input text."""
    lines = input_text.strip().split('\n')
    horizontal, vertical = parse_input(lines)

    print(horizontal)
    print(vertical)

    # Initialize empty grid
    grid = [[0] * 9 for _ in range(9)]
    
    if solve(grid, horizontal, vertical, verbose):
        return grid
    else:
        return None


def solve_from_stdin():
    """Solve enhanced Sudoku reading from standard input."""
    import sys
    lines = [line.rstrip('\n') for line in sys.stdin]
    solution = solve_enhanced_sudoku('\n'.join(lines))
    print_solution(solution)


def show_demo():
    """Simple demo showing how the solver works"""
    test_input = """ > <   < <   > <
v ^ v v ^ v ^ ^ v
 < <   < >   < <
v ^ v ^ v v ^ ^ v
 < <   < <   > >
 < >   > >   < >
v v ^ ^ v ^ ^ v v
 < >   > <   > >
^ v v v ^ v v ^ v
 > <   < >   > >
 < >   > >   > <
v v v v ^ ^ ^ ^ ^
 > <   < <   < <
^ ^ ^ ^ ^ v v v ^
 > >   < >   < <"""
    
    print("Enhanced Sudoku Solver Demo")
    print("=" * 40)
    print("Solving puzzle with constraints...")
    
    solution = solve_enhanced_sudoku(test_input, verbose=True)
    
    if solution:
        print("\nFinal Solution:")
        print_solution(solution)
    else:
        print("No solution found!")


def main():
    test_input = """ > <   < <   > <
v ^ v v ^ v ^ ^ v
 < <   < >   < <
v ^ v ^ v v ^ ^ v
 < <   < <   > >
 < >   > >   < >
v v ^ ^ v ^ ^ v v
 < >   > <   > >
^ v v v ^ v v ^ v
 > <   < >   > >
 < >   > >   > <
v v v v ^ ^ ^ ^ ^
 > <   < <   < <
^ ^ ^ ^ ^ v v v ^
 > >   < >   < <"""

    print("Enhanced Sudoku Solver")
    print("=" * 30)
    
    solution = solve_enhanced_sudoku(test_input)
    
    if solution:
        print("Solution:")
        print_solution(solution)
        print("\nTo see solving steps: show_demo()")
    else:
        print("No solution found!")

    show_demo()


if __name__ == "__main__":
    main()

File: c/futoshiki.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define VERBOSE_LIMIT 10
#define GRID_SIZE 9
#define H_CONSTRAINTS 6
#define V_CONSTRAINTS 6

typedef struct {
    char horizontal[GRID_SIZE][H_CONSTRAINTS];
    char vertical[V_CONSTRAINTS][GRID_SIZE];
} Constraints;

typedef struct {
    int grid[GRID_SIZE][GRID_SIZE];
    Constraints constraints;
    int step_count;
} FutoshikiSolver;

void init_solver(FutoshikiSolver* solver) {
    int i, j;
    
    // Initialize grid with zeros
    for (i = 0; i < GRID_SIZE; i++) {
        for (j = 0; j < GRID_SIZE; j++) {
            solver->grid[i][j] = 0;
        }
    }
    
    // Initialize constraints with null characters
    for (i = 0; i < GRID_SIZE; i++) {
        for (j = 0; j < H_CONSTRAINTS; j++) {
            solver->constraints.horizontal[i][j] = '\0';
        }
    }
    
    for (i = 0; i < V_CONSTRAINTS; i++) {
        for (j = 0; j < GRID_SIZE; j++) {
            solver->constraints.vertical[i][j] = '\0';
        }
    }
    
    solver->step_count = 0;
}

void parse_input(FutoshikiSolver* solver, const char* input) {
    // Simple parsing - this would need to be implemented based on your input format
    // For now, just initialize with some test constraints
    solver->constraints.horizontal[0][0] = '>';
    solver->constraints.horizontal[0][1] = '<';
    solver->constraints.vertical[0][0] = 'v';
    solver->constraints.vertical[0][1] = '^';
}

int is_valid_sudoku(FutoshikiSolver* solver, int row, int col, int val) {
    int i, j;
    
    // Check row constraint
    for (i = 0; i < GRID_SIZE; i++) {
        if (solver->grid[row][i] == val) {
            return 0;
        }
    }
    
    // Check column constraint
    for (i = 0; i < GRID_SIZE; i++) {
        if (solver->grid[i][col] == val) {
            return 0;
        }
    }
    
    // Check 3x3 subgrid constraint
    int start_row = (row / 3) * 3;
    int start_col = (col / 3) * 3;
    for (i = start_row; i < start_row + 3; i++) {
        for (j = start_col; j < start_col + 3; j++) {
            if (solver->grid[i][j] == val) {
                return 0;
            }
        }
    }
    
    return 1;
}

int is_valid_constraints(FutoshikiSolver* solver, int row, int col, int val) {
    // Simplified constraint checking - implement full logic as needed
    return 1;
}

int is_valid(FutoshikiSolver* solver, int row, int col, int val) {
    return is_valid_sudoku(solver, row, col, val) && 
           is_valid_constraints(solver, row, col, val);
}

int find_empty(FutoshikiSolver* solver, int* row, int* col) {
    int i, j;
    for (i = 0; i < GRID_SIZE; i++) {
        for (j = 0; j < GRID_SIZE; j++) {
            if (solver->grid[i][j] == 0) {
                *row = i;
                *col = j;
                return 1;
            }
        }
    }
    return 0;
}

int solve_recursive(FutoshikiSolver* solver, int verbose) {
    int row, col, val;
    
    solver->step_count++;
    
    if (verbose && solver->step_count <= VERBOSE_LIMIT) {
        printf("Step %d: Looking for empty position...\n", solver->step_count);
    }
    
    if (!find_empty(solver, &row, &col)) {
        return 1; // All positions filled
    }
    
    for (val = 1; val <= 9; val++) {
        if (is_valid(solver, row, col, val)) {
            solver->grid[row][col] = val;
            
            if (verbose && solver->step_count <= VERBOSE_LIMIT) {
                printf("  Placed %d at (%d, %d)\n", val, row, col);
            }
            
            if (solve_recursive(solver, verbose)) {
                return 1;
            }
            
            solver->grid[row][col] = 0; // Backtrack
        }
    }
    
    return 0;
}

// Restored C implementation for Futoshiki solver
int solve(FutoshikiSolver* solver, int verbose) {
    solver->step_count = 0;
    return solve_recursive(solver, verbose);
}

void print_solution(FutoshikiSolver* solver) {
    int i, j;
    for (i = 0; i < GRID_SIZE; i++) {
        for (j = 0; j < GRID_SIZE; j++) {
            printf("%d", solver->grid[i][j]);
            if (j < GRID_SIZE - 1) printf(" ");
        }
        printf("\n");
    }
}

void show_demo() {
    printf("Enhanced Sudoku Solver Demo (C Version)\n");
    printf("========================================\n");
    printf("Solving simplified puzzle...\n");
    
    FutoshikiSolver solver;
    init_solver(&solver);
    
    // Set up a simple test case
    solver.grid[0][0] = 1;
    solver.grid[0][1] = 2;
    solver.grid[1][0] = 3;
    
    if (solve(&solver, 1)) {
        printf("\nFinal Solution:\n");
        print_solution(&solver);
    } else {
        printf("No solution found!\n");
    }
}

int main() {
    printf("Enhanced Sudoku Solver (C Version)\n");
    printf("==================================\n");
    
    FutoshikiSolver solver;
    init_solver(&solver);
    
    // Simple test
    if (solve(&solver, 0)) {
        printf("Solution:\n");
        print_solution(&solver);
    } else {
        printf("No solution found!\n");
    }
    
    printf("\nRunning demo...\n");
    show_demo();
    
    return 0;
}
